jlm/llvm/ir/operators/MemoryStateOperations.cpp

```diff
@@ -29,6 +30,86 @@ MemoryStateMergeOperation::copy() const
   return std::make_unique<MemoryStateMergeOperation>(*this);
 }
 
+std::optional<std::vector<rvsdg::Output *>>
+MemoryStateMergeOperation::NormalizeSingleOperand(
+    const MemoryStateMergeOperation &,
+    const std::vector<rvsdg::Output *> & operands)
+{
+  if (operands.size() == 1)
+    return operands;
+
+  return std::nullopt;
+}
+
+std::optional<std::vector<rvsdg::Output *>>
+MemoryStateMergeOperation::NormalizeDuplicateOperands(
+    const MemoryStateMergeOperation &,
+    const std::vector<rvsdg::Output *> & operands)
+{
+  const util::HashSet<rvsdg::Output *> uniqueOperands(operands.begin(), operands.end());
```

Comment:
This operation will shuffle the operands when it is applied. Would it be worth it to use a vector in combination?

For the future maybe we could implement a insertion order maintaining hash set

Reply:
@haved Well, I need a set in order to determine in linear time whether we have duplicates. This would not work with a vector (with the same runtime). Is it a problem that this would shuffle the operands? There is no order defined on those operands anyway. Semantically, there is no problem either.

Reply:
@phate yes, I will just make a mental note about this being a place that introduces non-deterministic passes